src/iter.rs

```diff
@@ -71,15 +92,31 @@ impl<'a, S: Unpin> Iterator for IterMut<'a, S> {
 
 impl<S: Unpin> ExactSizeIterator for IterMut<'_, S> {}
 
-impl<'a, S> Iterator for IterPinRef<'a, S> {
-    type Item = Pin<&'a S>;
+impl<'a, S: Unpin> Iterator for IterMutWithToken<'a, S> {
+    type Item = (&'a mut S, usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next().map(|(stream, id)| (Pin::get_mut(stream), id))
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl<S: Unpin> ExactSizeIterator for IterMutWithToken<'_, S> {}
 
-    fn next(&mut self) -> Option<Pin<&'a S>> {
+impl<'a, S> Iterator for IterWithToken<'a, S> {
+    type Item = (&'a S, usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
         if self.task.is_null() {
             return None;
         }
         unsafe {
-            let stream = (*(*self.task).stream.get()).as_ref().unwrap();
+            let task = &*self.task;
+            let id = (*task).id;
+            let stream = (*task.stream.get()).as_ref().unwrap();
```

Comment:
Let's do the same thing here as we changed with `iter_mut`, which is to _not_ construct a reference inside `*self.task`. That _might_ not be okay later in the same function. Instead, just repeat `*self.task` in the two places it's used.